File: macos/QMK Toolbox/QMKWindow.m
#import "QMKWindow.h"

#import "AppDelegate.h"

@implementation QMKWindow

- (void)awakeFromNib {
    [self registerForDraggedTypes:[NSArray arrayWithObject:NSPasteboardTypeFileURL]];
}

- (NSDragOperation)draggingEntered:(id<NSDraggingInfo>)sender {
    NSPasteboard *pboard = [sender draggingPasteboard];

    if ([[pboard pasteboardItems] count] == 1) {
        NSString *fileExtension = [[NSURL URLFromPasteboard:pboard] pathExtension];

        if ([fileExtension isEqualToString:@"hex"] || [fileExtension isEqualToString:@"bin"]) {
            return NSDragOperationCopy;
        }
    }

    return NSDragOperationNone;
}

- (BOOL)performDragOperation:(id<NSDraggingInfo>)sender {
    NSPasteboard *pboard = [sender draggingPasteboard];
    [(AppDelegate *)[[NSApplication sharedApplication] delegate] setFilePath:[NSURL URLFromPasteboard:pboard]];
    return YES;
}

@end

File: macos/QMK Toolbox/QMKWindow.swift
import Foundation
import AppKit

class QMKWindow: NSWindow, NSDraggingDestination {
    override func awakeFromNib() {
        self.registerForDraggedTypes([.fileURL])
    }

    func draggingEntered(_ sender: NSDraggingInfo) -> NSDragOperation {
        let pasteboard = sender.draggingPasteboard
        if pasteboard.pasteboardItems?.count == 1 {
            if ["bin", "hex"].contains(NSURL(from: pasteboard)?.pathExtension) {
                return .copy
            }
        }
        return []
    }

    func performDragOperation(_ sender: NSDraggingInfo) -> Bool {
        let pasteboard = sender.draggingPasteboard
        guard let file = (NSURL(from: pasteboard) as URL?) else { return false }
        (NSApplication.shared.delegate as! AppDelegate).setFilePath(file)
        return true
    }
}
